test/main_test.go

```diff
@@ -0,0 +1,144 @@
+			expectApplyError: false,
+			vars: map[string]interface{}{
+				"name_prefix":         "",
+				"ingress_cidr_blocks": []string{"0.0.0.0/0"},
+				"egress_cidr_blocks":  []string{"0.0.0.0/0"},
+				"license_key":         "averysecretkey",
+				"tags":                make(map[string]string),
+				"emr_tags":            make(map[string]string),
+				"emr_abac_valid_tags": make(map[string][]string),
+			},
+		},
+		{
+			testName:         "TestRootModuleYaml",
+			tfDir:            "test_examples/root_module",
+			expectApplyError: false,
+			vars: map[string]interface{}{
+				"name_prefix": "",
+			},
+		},
+	}
+}
+
+func TestAllCases(t *testing.T) {
+	// os.Setenv("TERRATEST_REGION", "us-east-1")
+
+	// os.Setenv("SKIP_pick_new_randoms", "true")
+	// os.Setenv("SKIP_setup_options", "true")
+	// os.Setenv("SKIP_create", "true")
+	// os.Setenv("SKIP_validate", "true")
+	// os.Setenv("SKIP_teardown", "true")
+
+	testCases := initTestCases()
+
+	for _, testCase := range testCases {
+		testCase := testCase
+
+		t.Run(testCase.testName, func(t *testing.T) {
+			t.Parallel()
+
+			tempTestFolder := test_structure.CopyTerraformFolderToTemp(t, "../", testCase.tfDir)
+
+			test_structure.RunTestStage(t, "pick_new_randoms", func() {
+
+				usRegions := []string{"us-east-1", "us-east-2", "us-west-1", "us-west-2"}
+				// This function will first check for the Env Var TERRATEST_REGION and return its value if it is set.
+				awsRegion := aws.GetRandomStableRegion(t, usRegions, nil)
+
+				test_structure.SaveString(t, tempTestFolder, "region", awsRegion)
+				// some resources require a lowercase alphabet as first char
+				test_structure.SaveString(t, tempTestFolder, "unique_id", fmt.Sprintf("a%s", strings.ToLower(random.UniqueId())))
+			})
+
+			defer test_structure.RunTestStage(t, "teardown", func() {
+				teraformOptions := test_structure.LoadTerraformOptions(t, tempTestFolder)
+				terraform.Destroy(t, teraformOptions)
+			})
+
+			test_structure.RunTestStage(t, "setup_options", func() {
+				awsRegion := test_structure.LoadString(t, tempTestFolder, "region")
+				uniqueID := test_structure.LoadString(t, tempTestFolder, "unique_id")
+
+				testCase.vars["name_prefix"] = uniqueID
+
+				terraformOptions := terraform.WithDefaultRetryableErrors(t, &terraform.Options{
+					TerraformDir: tempTestFolder,
+					Vars:         testCase.vars,
+					EnvVars: map[string]string{
+						"AWS_DEFAULT_REGION": awsRegion,
+					},
+				})
+
+				test_structure.SaveTerraformOptions(t, tempTestFolder, terraformOptions)
+			})
+
+			test_structure.RunTestStage(t, "create", func() {
+				terraformOptions := test_structure.LoadTerraformOptions(t, tempTestFolder)
+
+				_, err := terraform.InitAndApplyE(t, terraformOptions)
+
+				if testCase.expectApplyError {
+					require.Error(t, err)
+					// If it failed as expected, we should skip the rest (validate function).
+					t.SkipNow()
+				}
+
+			})
+
+			test_structure.RunTestStage(t, "validate", func() {
```

Comment:
Structurally in the other modules, we put the validation logic into a different file to keep it separate from this file, which is mostly testing boilerplate.

Reply:
Fixed

Reply:
A great start! Could go a bit further and get even more into that utils class.

Like if you look to the EMR module for example, it delegates all of the functionality to the `validateModuleOutputs` method:
https://github.com/Datatamer/terraform-aws-emr/blob/master/test/create_examples_emr_test.go#L125-L131

Here we're still doing quite a bit of yaml parsing, etc., which I think would be better suited to happen in a dedicated method.

Motivation is to keep this file fairly boring, and make the other one have all the interesting case logic. We have a ticket in the backlog to do eventually build out a shared library, and this file would be a great place to start since it's 90% the same across the modules.